x509/src/certificate_document.rs

```diff
@@ -0,0 +1,166 @@
+//! CertificateDocument implementation
+
+use crate::Certificate;
+use der::{Error, Result};
+
+use alloc::vec::Vec;
+use core::fmt;
+use der::{Decodable, Document};
+
+#[cfg(feature = "std")]
+use std::path::Path;
+
+#[cfg(feature = "pem")]
+use {
+    alloc::string::String,
+    core::str::FromStr,
+    der::pem::{self, LineEnding},
+};
+
+/// Certificate document.
+///
+/// This type provides storage for [`Certificate`] encoded as ASN.1
+/// DER with the invariant that the contained-document is "well-formed", i.e.
+/// it will parse successfully according to this crate's parsing rules.
+#[derive(Clone)]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+pub struct CertificateDocument(Vec<u8>);
+
+impl<'a> TryFrom<&'a [u8]> for Certificate<'a> {
+    type Error = Error;
+
+    fn try_from(bytes: &'a [u8]) -> Result<Self> {
+        Self::from_der(bytes)
+    }
+}
+
+impl<'a> Document<'a> for CertificateDocument {
+    type Message = Certificate<'a>;
+    const SENSITIVE: bool = false;
+}
+
+impl CertificateDocument {
+    /// Parse [`Certificate`] from buffer containing DER-encoded certificate
+    pub fn from_certificate_der(bytes: &[u8]) -> Result<Self> {
```

Comment:
Since these aren't trait methods and they're defined on `CertificateDocument` directly, I think having `*_certificate_*` in the name is redundant

Reply:
Actually they can just go away. Tests pass when directly leaning on blanket impl.

Reply:
Aah yeah, indeed!